### PR TITLE
Staking V3

### PR DESCRIPTION
**Pull request by AurevoirXavier**
- [x] Remove all unused storage
- [x] Simplify election & reward logic
- [x] Use `on_idle` to distribute reward in background
- [x] Use `on_idle` to migrate un-migrated accounts in background
- [x] Benchmark
- [x] Migration

Closes:
- #1606 
- #1609

### CODE REVIEW

File: pallet/ethtx-forwarder/src/lib.rs [hunk 1]
```diff
@@ -280,7 +280,11 @@ impl<T: Config> Pallet<T> {
 		.validate_in_block_for(&who)
 		.and_then(|v| v.with_base_fee())
 		.and_then(|v| v.with_balance_for(&who))
-		.map_err(|e| <Error<T>>::ValidationError(e))?;
+		.map_err(|e| {
+			log::error!("transaction validation failed due to {e:?}");
```
[the remark this answers is not in the document]
Comment (AurevoirXavier) [reply pallet/ethtx-forwarder/src/lib.rs:284]: log crate will include the filename automatically.

File: pallet/staking/src/lib.rs [hunk 2]
```diff
@@ -163,125 +89,45 @@ pub mod pallet {
...
-	#[pallet::storage]
-	#[pallet::getter(fn nominator_of)]
-	pub type Nominators<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, T::AccountId>;
-
 	/// Number of blocks authored by the collator within current session.
 	#[pallet::storage]
 	#[pallet::unbounded]
 	#[pallet::getter(fn authored_block_count)]
-	pub type AuthoredBlocksCount<T: Config> =
+	pub type AuthoredBlockCount<T: Config> =
```
Comment (hackfisher) [inline pallet/staking/src/lib.rs:130]: Need to note migration for this storage specifically.

File: pallet/deposit/src/lib.rs [hunk 3]
```diff
@@ -159,204 +116,73 @@ pub mod pallet {
...
-						to_claim.1.push(d.id);
 
-						false
+			#[cfg(test)]
+			let wt = Weight::zero().add_ref_time(10);
+			#[cfg(not(test))]
+			let wt = <T as Config>::WeightInfo::migrate_for();
+			let mut ds_to_migrate = BTreeMap::<T::AccountId, (Vec<Deposit>, usize)>::new();
+
+			'outer: for (w, ds) in <Deposits<T>>::iter() {
```
Comment (AurevoirXavier) [inline pallet/deposit/src/lib.rs:136]: Migrate 10 deposits every block.

File: pallet/account-migration/src/lib.rs [hunk 4]
```diff
@@ -108,10 +108,7 @@ pub mod pallet {
 		> + pallet_assets::Config<Balance = Balance, AssetId = AssetId>
 		+ pallet_balances::Config<Balance = Balance>
 		+ darwinia_deposit::Config
-		+ darwinia_staking::Config<
```
Comment (AurevoirXavier) [inline pallet/account-migration/src/lib.rs:111]: Skip inserting ledger records.

File: pallet/staking/src/lib.rs [hunk 5]
```diff
@@ -371,231 +177,49 @@ pub mod pallet {
...
-					.fold(Weight::zero(), |acc, c| acc
-						+ Self::payout_inner(c).unwrap_or(Zero::zero())))
-				.unwrap_or_default();
-
-				w1 + w2
-			} else {
-				Zero::zero()
-			}
+		fn on_idle(_: BlockNumberFor<T>, mut remaining_weight: Weight) -> Weight {
+			Self::idle_allocate_ring_staking_reward(&mut remaining_weight);
```
Comment (AurevoirXavier) [inline pallet/staking/src/lib.rs:189]: limit reward count

**Pull request merged**